dataprocspawner/spawner.py

```diff
@@ -1167,22 +1180,36 @@ def _apply_users_configs(self, cluster_data):
           }
       )
 
+    if self.user_options.get('init_actions'):
+      for init_action in self.user_options.get('init_actions').split(','):
+        if init_action.startswith('gs://'):
+          config['initialization_actions'].append(
+            {
+              'executable_file': init_action
+            }
+          )
+
+    if self.user_options.get('image_version'):
+      if self.user_options.get('custom_image'):
+        config['master_config']['image_uri'] = self.user_options.get('custom_image')
+        config['worker_config']['image_uri'] = self.user_options.get('custom_image')
+        config['software_config']['image_version'] = \
```

Comment:
Is this necessary? Dataproc's frontend should do the work of inferring the image version given a custom image URI, there's no reason for us to validate it here as well.

Reply:
setting up the image version here is important if admins define different softwareConfig.imageVersion in YAML file.
Example: if 1.5-debian10 is set in YAML and custom image (1.5.17-debian10) is selected in UI, you will get an exception:
```
Error: HTTP 500: Internal Server Error (Error in Authenticator.pre_spawn_start:
InvalidArgument 400 Specified image version '1.5-debian10' doesn't match implied version '1.5.17-debian10' from uri 'https://www.googleapis.com/compute/v1/projects/<project-id>/global/images/<image-id>'!)
```
Additionally we check image_version for Component Gateway compatibility, so it has to be set up properly before the validation occurs.

Reply:
We can unset the image version if a custom image is set and vice versa.

The point about image version checking for Component Gateway is well taken. I'm a little concerned about latency implications for a call to GCE before spawning, but let's merge this and see if it actually has an impact.